Fulltext addon for bloom

ragadeeshu opened the pull request.

ragadeeshu committed:
Initial spike for the insight index.

Writing and reading the index almost works.

ragadeeshu committed:
Properly update readers

ragadeeshu committed:
Actually tokenize strings

ragadeeshu committed:
Checkstyle fixes insight index

ragadeeshu committed:
Cleanup of InsightFieldEncoding

ragadeeshu committed:
Properly close bloom index resources

ragadeeshu committed:
Cleanup junk files from insight index

ragadeeshu committed:
Insight index added support for removal of nodes

ragadeeshu committed:
Insight index support multiple properties per node

ragadeeshu committed:
Fix compilation errors after rebase

ragadeeshu committed:
Support property changes in insight index

ragadeeshu committed:
Initial support for rels in insight index

ragadeeshu committed:
Fully separate node and relationship insight index

ragadeeshu committed:
Bloom index order results after lucene score

ragadeeshu committed:
Only index indexed properties in bloom index

ragadeeshu committed:
Rename from insight to bloom index

ragadeeshu committed:
Extract bloom index to separate module.

ragadeeshu committed:
Introduce setting for bloom indexed properties

ragadeeshu committed:
Preparations for configurable bloom index analyzer

ragadeeshu committed:
Change to StandardAnalyzer in bloom index

ragadeeshu committed:
Add bloom kernel extensions

ragadeeshu committed:
Add procedures for querying the bloom index

ragadeeshu committed:
Fix dependencies for bloom index

ragadeeshu committed:
Add support for all property types for blomm index

ragadeeshu committed:
Bloom index now properly indexes relationships

ragadeeshu committed:
Fix notice for bloom index

ragadeeshu committed:
Configurable analyzer for bloom index

burqen commented on enterprise/bloom-index/pom.xml:
```diff
@@ -0,0 +1,149 @@
+    </license>
+  </licenses>
+
+  <dependencies>
```
Please divide "main" and test dependencies with `<!--Test dependencies-->`.

burqen commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -409,6 +409,11 @@
+    // Bloom index
+    @Internal
+    public static final Setting<List<String>> bloom_indexed_properties =
+            setting( "unsupported.dbms.bloom_indexed_properties", STRING_LIST, TRUE );
```
What does `TRUE` mean for default value here? Other `Setting<List<String>>` settings use an empty string as default value.

burqen commented on enterprise/bloom-index/LICENSES.txt:
```diff
@@ -0,0 +1,247 @@
+This file contains the full license text of the included third party
```
`mvn clean install` gives
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:39 min (Wall Clock)
[INFO] Finished at: 2017-08-29T17:04:40+02:00
[INFO] Final Memory: 148M/1570M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.neo4j.build.plugins:licensing-maven-plugin:1.7.5:check (list-all-licenses) on project neo4j-bloom-index: Generated file differs from the existing file.
[ERROR] Generated: /Users/antonpersson/neo4j/neo4j-3.3/enterprise/bloom-index/target/neo4j-bloom-index-3.3.0-SNAPSHOT-LICENSES.txt
[ERROR] Existing: /Users/antonpersson/neo4j/neo4j-3.3/enterprise/bloom-index/target/../LICENSES.txt
```

Easy to fix by overriding existing file with generated one.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomKernelExtension.java:
```diff
@@ -0,0 +1,72 @@
+    private final File storeDir;
+    private final Config config;
+    private final FileSystemAbstraction fileSystemAbstraction;
+    private GraphDatabaseService db;
```
GraphDatabaseService can be final

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomKernelExtension.java:
```diff
@@ -0,0 +1,72 @@
+    private final Config config;
+    private final FileSystemAbstraction fileSystemAbstraction;
+    private GraphDatabaseService db;
+    private Procedures procedures;
```
Procedures can be final

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomKernelExtension.java:
```diff
@@ -0,0 +1,72 @@
+    public void init() throws IOException, ProcedureException
+    {
+        bloomIndex = new BloomIndex( fileSystemAbstraction, storeDir, config );
+        bloomIndexTransactionEventUpdater = bloomIndex.getUpdater();
```
I would prefer if we will split direct relationship between index and updater and instead have some sort of index provider that aware about all existent indexes, that updater can use to perform index update.
By doing that we will allow update to update any number of indexes that we will have in future

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomNodeProcedure.java:
```diff
@@ -0,0 +1,78 @@
+        }
+        catch ( Exception e )
+        {
+            throw new ProcedureException( Status.statusCodeOf( e ), e, "Failed to query bloom index for nodes", input );
```
Status.statusCodeOf can return null, I think we should put status to be at least `Status.Procedure.ProcedureCallFailed` in case of exception will occur

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomRelationshipProcedure.java:
```diff
@@ -0,0 +1,78 @@
+    private static final String OUTPUT_NAME = "relid";
+    private BloomIndex bloomIndex;
+
+    BloomRelationshipProcedure( BloomIndex bloomIndex )
```
2 introduced procedures are almost the same would be better if we will introduce generic index reader procedure that can query supplier index reader

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomDocumentStructure.java:
```diff
@@ -0,0 +1,124 @@
+        return document.document;
+    }
+
+    public static long getNodeId( Document from )
```
this method is never used, plus that is not really true that that field represents node id right?

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomDocumentStructure.java:
```diff
@@ -0,0 +1,124 @@
+
+        private DocWithId()
+        {
+            idField = new StringField( ID_KEY, "", YES );
```
since we never get value from that field do we need it to be stored?

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomNodeProcedure.java:
```diff
@@ -0,0 +1,78 @@
+public class BloomNodeProcedure extends CallableProcedure.BasicProcedure
+{
+    private static final String PROCEDURE_NAME = "bloomNodes";
+    private static final String[] PROCEDURE_NAMESPACE = {"dbms", "bloom"};
```
The name space for updating and querying the index should be `db` because it manages the data. Creating the index in the first place should be `dbms` because it manages the database.

When it comes to the `bloom` part I'm not sure that is the namespace that we want. However I don't have any though through suggestions right now. Let's discuss it.

burqen replied:
Also, the `PROCEDURE_NAMESPACE` should be shared between `BloomNodeProcedure` and `BloomRelationshipProcedure`.

jexp replied:
Was there a reason for not using the public procedure API ? /cc @craigtaverner

ragadeeshu replied:
@jexp this seemed like the easiest way :)

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomRelationshipProcedure.java:
```diff
@@ -0,0 +1,78 @@
+public class BloomRelationshipProcedure extends CallableProcedure.BasicProcedure
+{
+    private static final String PROCEDURE_NAME = "bloomRelationships";
+    private static final String[] PROCEDURE_NAMESPACE = {"dbms", "bloom"};
```
See namespace comment above.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndex.java:
```diff
@@ -0,0 +1,89 @@
+    {
+        this.properties = config.get( GraphDatabaseSettings.bloom_indexed_properties ).toArray( new String[0] );
+        Analyzer analyzer = new StandardAnalyzer();
+        Factory<IndexWriterConfig> population = () -> IndexWriterConfigs.population( analyzer );
```
do we need configuration to be the one that used for the population in the product?

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndex.java:
```diff
@@ -0,0 +1,89 @@
+        Factory<IndexWriterConfig> population = () -> IndexWriterConfigs.population( analyzer );
+        WritableIndexPartitionFactory partitionFactory = new WritableIndexPartitionFactory( population );
+
+        LuceneIndexStorageBuilder storageBuilder = LuceneIndexStorageBuilder.create();
```
i would extract all of the index building into specialized index builder

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndex.java:
```diff
@@ -0,0 +1,89 @@
+    private final BloomLuceneIndex relationshipIndex;
+    private final String[] properties;
+
+    public BloomIndex( FileSystemAbstraction fileSystem, File file, Config config ) throws IOException
```
even if for now idea of grouping indexes will work, i think in a long term we will need node/relationship indexes to be independent with independent updaters as well, and I would even prefer that now compare to a current index grouping

burqen replied:
`storeDir` would be a more descriptive name for `file`

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+        Map<Long,Map<String,Object>> relationshipMap = ((Map<Long,Map<String,Object>>[]) state)[1];
+        updatePropertyData( data.removedRelationshipProperties(), relationshipMap, relationshipIndex );
+        updatePropertyData( data.assignedRelationshipProperties(), relationshipMap, relationshipIndex );
+        deleteIndexData( data.deletedNodes(), nodeIndex );
```
this line needs to be updated to handle relationship index instead

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+    private <E extends Entity> void updatePropertyData( Iterable<PropertyEntry<E>> propertyEntries, Map<Long,Map<String,Object>> state,
+            WritableDatabaseBloomIndex nodeIndex )
+    {
+        for ( PropertyEntry<E> propertyEntry : propertyEntries )
```
it looks like we do not really need go over `propertyEntries`, instead we can go state since it's already should contain all the update that we need to do, as a nice benefit by doing that we will not need to perform additional map lookup

ragadeeshu replied:
Reacting to removed properties is much harder without iterating over propertyEntries. But perhaps the method could be split for additions and removals.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+        }
+        try
+        {
+            nodeIndex.maybeRefreshBlocking();
```
why do we do this? we still eventual right?

ragadeeshu replied:
If nothing else, this is needed for the tests, at least the way it looks now. I've extracted it for clarity and it is certainly something we can revisit.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+            }
+            catch ( IOException e )
+            {
+                e.printStackTrace();
```
looks like a nice ability to handle exceptions in a better way

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+
+    private void deleteIndexData( Iterable<Node> nodes, WritableDatabaseBloomIndex nodeIndex )
+    {
+        for ( Node node : nodes )
```
same comments here as above

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+            }
+        } );
+        data.assignedRelationshipProperties().forEach(
+                propertyEntry -> relationshipMap.put( propertyEntry.entity().getId(), propertyEntry.entity().getProperties( properties ) ) );
```
can we have a case where node or relationship map will be updated twice? (its part of removedRelationshipProperties and assignedRelationshipProperties,  for example)

ragadeeshu replied:
Yes, but it will still end up with the correct data since it fetches the current properties from the database.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndex.java:
```diff
@@ -0,0 +1,144 @@
+        this.analyzer = analyzer;
+    }
+
+    private static final String KEY_STATUS = "status";
```
static fields/ fields should be declared before constructor

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndex.java:
```diff
@@ -0,0 +1,144 @@
+    private String[] properties;
+    private final Analyzer analyzer;
+
+    BloomLuceneIndex( PartitionedIndexStorage indexStorage, IndexPartitionFactory partitionFactory, String[] properties, Analyzer analyzer )
```
It looks like most of it is the same as LuceneSchemaIndex (except uniqueness and maybe readers) maybe we can extract inherit that functionality from it instead?

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/PartitionedInsightBloomWriter.java:
```diff
@@ -0,0 +1,132 @@
+ * {@link #MAXIMUM_PARTITION_SIZE}.
+ * First observable partition that satisfy writer criteria is used for writing.
+ */
+class PartitionedInsightBloomWriter
```
this writer is almost a copy of PartitionedIndexWriter, why do we need to create it why not reuse it for this indexes as well?

burqen replied:
Nice with javadoc but could be updated to not talk about schema index.

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/SimpleBloomIndexReader.java:
```diff
@@ -0,0 +1,103 @@
+{
+    private final PartitionSearcher partitionSearcher;
+    private final Analyzer analyzer;
+    private String[] properties;
```
final properties

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/SimpleBloomIndexReader.java:
```diff
@@ -0,0 +1,103 @@
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        QueryParser multiFieldQueryParser = new MultiFieldQueryParser( properties, analyzer );
+        multiFieldQueryParser.setDefaultOperator( QueryParser.Operator.OR );
+        for ( String s : query )
```
boolean query can have max of 1024 clauses with this two loops it looks like you can over exceed it quite easily, can we somehow avoid doing it? (query time will also depend from how complex query you will end-up with)

ragadeeshu replied:
I have not found another way yet, at the very least. The number of clauses is `(number of properties in index * number of strings to search for)`

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/WritableDatabaseBloomIndex.java:
```diff
@@ -0,0 +1,80 @@
+
+class WritableDatabaseBloomIndex extends WritableAbstractDatabaseIndex<BloomLuceneIndex>
+{
+    private BloomLuceneIndex bloomLuceneIndex;
```
can be final field

MishaDemianenko commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/WritableDatabaseBloomIndex.java:
```diff
@@ -0,0 +1,80 @@
+        return bloomLuceneIndex.getIndexWriter( this );
+    }
+
+    public BloomIndexReader getIndexReader() throws IOException
```
most of methods here are not used

MishaDemianenko commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/integration/BloomIT.java:
```diff
@@ -0,0 +1,91 @@
+    public static final String NODEID = "nodeid";
+    public static final String RELID = "relid";
+    @Rule
+    public final EphemeralFileSystemRule fs = new EphemeralFileSystemRule();
```
quite strange that IT test operate on top of ephemeral file system

MishaDemianenko commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndexUpdaterTest.java:
```diff
@@ -0,0 +1,481 @@
+                PrimitiveLongIterator one = reader.query( "1" );
+                assertEquals( firstID, one.next() );
+                assertFalse( one.hasNext() );
+                PrimitiveLongIterator twohundredthirtyfour = reader.query( "234" );
```
thisvariablenamecanbeabitmorereadable:)

MishaDemianenko commented on enterprise/bloom-index/pom.xml:
```diff
@@ -0,0 +1,149 @@
+  <version>3.3.0-SNAPSHOT</version>
+
+  <name>Neo4j - Bloom index</name>
+  <description>Enterprise locks deferred to commit time</description>
```
this pom file needs a bit more attention

MishaDemianenko commented on enterprise/bloom-index/pom.xml:
```diff
@@ -0,0 +1,149 @@
+  <packaging>jar</packaging>
+
+  <scm>
+    <url>https://github.com/neo4j/neo4j/tree/3.0/enterprise/deferred-locks</url>
```
here as well

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -409,6 +409,11 @@
 
+    // Bloom index
+    @Internal
+    public static final Setting<List<String>> bloom_indexed_properties =
```
we should not expose extension settings in global database settings file, please introduce plugin local setting file instead

MishaDemianenko reviewed (changes requested):
First round of review completed, have some questions here and there, not sure in general about namespace for procedures and name in general. Also for this amount of code would expect to see more tests

burqen commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/integration/BloomIT.java:
```diff
@@ -0,0 +1,91 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class BloomIT
```
Suggestions for additional stuff to test here:

Individual tests:
`shouldNotReturnNodesThatAreNotIndexHits`
`shouldNotReturnRelationshipsThatAreNotIndexHits`
`somethingAboutCaseSensitivity` (should or shouldNot ignore case?)
`shouldQueryForSuperLongString`
`partialHitShouldNotBeAHit` ( someNode:Label {prop="This is my property"}), query for "This is ma property" (or something like this)
`shouldQueryForSpecialCharacters` (which characters are supported / not supported?)
`shouldNotHitOnRemovedProperties`

Additional test areas:
- What happens when you shutdown, update config and start up again? Test for adding and removing properties from config.
- Adding kernel extension on existing store with data relevant for bloom index?
- What happens if index is corrupt, missing file for example. How should / will it behave?
- General randomised tests and randomised concurrency tests would also be great to shake out some bugs.

burqen replied:
Feel free to distribute those tests outside of `BloomIT`, should you see fit.

ragadeeshu replied:
Tests relating to restarts, config changes and other such shenanigans are out of scope for this PR :)

ragadeeshu replied:
Case sensitivity is up to the analyzer you decide to use, the same goes for special characters.

burqen commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndexUpdaterTest.java:
```diff
@@ -0,0 +1,481 @@
+                Node node2 = db.createNode( LABEL );
+                secondID = node2.getId();
+                node2.setProperty( "prop", "A zebroid (also zedonk, zorse, zebra mule, zonkey, and zebmule) is the offspring of any cross " +
+                        "between a zebra and any other equine: essentially, a zebra hybrid." );
```
The more you know 🎉

burqen commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndexUpdaterTest.java:
```diff
@@ -0,0 +1,481 @@
+                assertFalse( one.hasNext() );
+                PrimitiveLongIterator twohundredthirtyfour = reader.query( "234" );
+                assertEquals( secondID, twohundredthirtyfour.next() );
+                assertFalse( twohundredthirtyfour.hasNext() );
```
Nice that you verify end of iterator! 👍

burqen commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndexUpdaterTest.java:
```diff
@@ -0,0 +1,481 @@
+                PrimitiveLongIterator strings = reader.query( "live" );
+                assertEquals( firstID, strings.next() );
+                assertFalse( strings.hasNext() );
+                PrimitiveLongIterator ints = reader.query( "27" );
```
Is it possible to query on multiple parts of the array? Like you would search on multiple words in a sentence.

burqen commented on enterprise/bloom-index/src/test/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndexUpdaterTest.java:
```diff
@@ -0,0 +1,481 @@
+            try ( BloomIndexReader reader = bloomIndex.getNodeReader() )
+            {
+
+                iterator = reader.query( "tomtar", "karl" );
```
Querying for "karl" will not give any additional hits because all the nodes will have a hit on "tomtar". Suggestion is to remove "tomtar" from the properties of `node2`

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/PartitionedBloomIndexReader.java:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @see SimpleBloomIndexReader
+ */
+class PartitionedBloomIndexReader implements BloomIndexReader
```
Would be very nice with some tests for this guy.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomNodeProcedure.java:
```diff
@@ -0,0 +1,78 @@
+
+    BloomNodeProcedure( BloomIndex bloomIndex )
+    {
+        super( procedureSignature( new QualifiedName( PROCEDURE_NAMESPACE, PROCEDURE_NAME ) ).in( "terms", Neo4jTypes.NTList( Neo4jTypes.NTString ) ).out(
```
Formatting

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/integration/BloomRelationshipProcedure.java:
```diff
@@ -0,0 +1,78 @@
+
+    BloomRelationshipProcedure( BloomIndex bloomIndex )
+    {
+        super( procedureSignature( new QualifiedName( PROCEDURE_NAMESPACE, PROCEDURE_NAME ) ).in( "terms", Neo4jTypes.NTList( Neo4jTypes.NTString ) ).out(
```
Formatting.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomDocumentStructure.java:
```diff
@@ -0,0 +1,124 @@
+        {
+            for ( Map.Entry<String,Object> entry : values.entrySet() )
+            {
+                Field field = encodeValueField( entry.getKey(), Values.of( entry.getValue() ) );
```
I think @fickludd made some efforts to remove `Values.of( object )` and instead use proper `Value` objects everywhere. The data flow here come from `beforeCommit` in `BloomIndexTransactionEventUpdater` and ultimately from `PropertyContainer.getProperties` (which I don't think we can change as it is part of public API).

@fickludd , do you have any insight into what could be done to avoid this (maybe unnecessary) `Values.of(object)`?

fickludd replied:
`Values.of()` is slow and should preferably not be used in hot paths. However, if you are feeding the index via the public API you have no choice but to convert Value -> Object -> Value. Some alternatives:
 * Store object in the bloom index if it's that separate from Neo4j
 * Pay the cost of the conversions
 * Create special internal API for feeding values to the new index

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndex.java:
```diff
@@ -0,0 +1,144 @@
+
+import static java.util.Collections.singletonMap;
+
+class BloomLuceneIndex extends AbstractLuceneIndex
```
From javadoc of `AbstractLuceneIndex`: 
```
 * Class and it's subclasses should not be directly used, instead please use corresponding writable or read only
 * wrapper.
 * @see WritableAbstractDatabaseIndex
 * @see ReadOnlyAbstractDatabaseIndex
```

Would it be suitable to use `WritableAbstractDatabaseIndex` for this guy?
Do we have any concept of "read only" for bloom index?

ragadeeshu replied:
The read-only counterpart you are looking for would be the `FulltextReader` interface.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndex.java:
```diff
@@ -0,0 +1,89 @@
+
+        LuceneIndexStorageBuilder storageBuilder = LuceneIndexStorageBuilder.create();
+        storageBuilder.withFileSystem( fileSystem ).withIndexIdentifier( "insightNodes" ).withDirectoryFactory(
+                directoryFactory( false, fileSystem ) ).withIndexRootFolder( Paths.get( file.getAbsolutePath(), "insightindex" ).toFile() );
```
"insightindex" and "insightNodes", as well as "insightindex" and "insightRelationships" below.

The conceptual naming (bloom-index, insight-index, something-else-index?) in this module is in general a bit "up in the air", which is understandable. Just don't miss this part when naming is decided. I can see the final name changing using some string searching functionality so please change this to something with "bloom" ;)

burqen replied:
The folder structure will look like this (including only folder related to indexes)
```
graph-db
├── index
├── insightindex
│   ├── insightNodes
│   │   └── 1
│   └── insightRelationships
│       └── 1
├── schema
.
.
.
```
("index" is really legacy / manual indexes)
Except for the naming (insight) i think this is a reasonable structure. I especially like that it is _not_ under schema or index seeing as it is a completely different thing.
🎉 
That being said, this folder structure will be hard to change when this is in production so we want to think long and hard about it.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomLuceneIndex.java:
```diff
@@ -0,0 +1,144 @@
+
+    private final TaskCoordinator taskCoordinator = new TaskCoordinator( 10, TimeUnit.MILLISECONDS );
+
+    public PartitionedInsightBloomWriter getIndexWriter( WritableDatabaseBloomIndex writableIndex ) throws IOException
```
The structure here is a bit (maybe unnecessarily?) confusing.

- `BloomLuceneIndex` is an `AbstractLuceneIndex`
- `WritableDatabaseBloomIndex` act as a wrapper for `BloomLuceneIndex` to present it as a `WritableAbstractDatabaseIndex` (subclass of `AbstractLuceneIndex`).
- But this method (`getIndexWriter`) in `BloomLuceneIndex` take `WritableDatabaseBloomIndex` as input. It takes it's own wrapper class as input.

ragadeeshu replied:
Sadly, I think that this is the way it is meant to be, at least for now.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/PartitionedInsightBloomWriter.java:
```diff
@@ -0,0 +1,132 @@
+
+    // by default we still keep a spare of 10% to the maximum partition size: During concurrent updates
+    // it could happen that 2 threads reserve space in a partition (without claiming it by doing addDocument):
+    private final Integer MAXIMUM_PARTITION_SIZE = Integer.getInteger( "luceneSchemaIndex.maxPartitionSize",
```
Do we want to use the same partition size as whoever is using this guy? If so, this should be extracted to some static variable.

burqen commented on enterprise/bloom-index/src/main/java/org/neo4j/kernel/api/impl/bloom/BloomIndexTransactionEventUpdater.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    @Override
+    public Object beforeCommit( TransactionData data ) throws Exception
```
Looks like we don't separate removed properties from updated properties. Do we have tests that verify that remove properties work?

burqen reviewed (changes requested):
First review round finished 🎉 
In general I think it looks good as a first approach. The integration with Neo utilises the kernel extension pattern in a clear way, which is nice!

There are quite some work to be done still though. In this PR or as future work.

Other than the things that I have commented on I would also like it if we could javadoc this module. Especially since it is built completely from scratch it is a great opportunity to be very rigid with the documentation and writing it now will be much easier than down the line. At least every class and public method should have documentation.
This would make me very happy 😄

ragadeeshu committed:
Introduce provider for fulltext helpers

This provider is then used by the updater, and support any number of
indexes, in theory at least.

ragadeeshu committed:
Fulltext addon cleanup

ragadeeshu committed:
Simplify the updater for the fulltext addon

ragadeeshu committed:
Fulltext addon removal of unused code + cleanup

ragadeeshu committed:
Fulltext addon cleanup

ragadeeshu committed:
Move bloom addon settings to separate file

ragadeeshu committed:
Tests for fulltext analyzer configuration

ragadeeshu committed:
Fulltext addon pom cleanup

ragadeeshu committed:
Fulltext addon fuzzy query support

ragadeeshu committed:
Reduce surface of fulltext addon

ragadeeshu committed:
javadoc for the fulltext addon

ragadeeshu committed:
Some additional tests for the fulltext addon

ragadeeshu committed:
Cleanup after fulltext addon rebase

ragadeeshu committed:
Fix licenses file for fulltext addon

ragadeeshu committed:
Javadoc and naming fixes for fulltext addon

ragadeeshu committed:
Added services file for bloom kernel extension

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomKernelExtension.java:
```diff
@@ -0,0 +1,84 @@
+import org.neo4j.kernel.impl.proc.Procedures;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+
+class BloomKernelExtension extends LifecycleAdapter
```
module was renamed to `Full text addon`, in the same time extension, properties and procedure still have bloom as part of their names. I think it's better if we will keep those names in sync and rename all of them.

ragadeeshu replied:
As discussed, will keep the bloom name for surface for the time being, but it may be subject to change.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomKernelExtension.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public void init() throws IOException, ProcedureException
+    {
+        String[] properties = config.get( LoadableBloomFulltextConfig.bloom_indexed_properties ).toArray( new String[0] );
```
here you convert list of properties to array. As soon as those will be passed into LuceneFulltext they will be converted back again to list

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomKernelExtension.java:
```diff
@@ -0,0 +1,84 @@
+        try
+        {
+            Class configuredAnalayzer = Class.forName( config.get( LoadableBloomFulltextConfig.bloom_analyzer ) );
+            analyzer = (Analyzer) configuredAnalayzer.newInstance();
```
analyzer should be extracted into a separate method

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomKernelExtension.java:
```diff
@@ -0,0 +1,84 @@
+            throw new RuntimeException( "Could not create the configured analyzer", e );
+        }
+
+        FulltextProvider provider = FulltextProvider.instance( db );
```
i can't understand why do you need FulltextProvider to be that harsh singleton with a lot of synch methods and added complexity around them. it would be much easier if we will create an instance of provider during init method call and then close it on shutdown if it's there. Nobody else will gonna do it and there will be no races if we will gonna do it that way.

ragadeeshu replied:
I can see a future where you would have one or more extensions (bloom, sync fulltext, async fulltext? maybe) that all want to make sure that there is an instance, and in such a case this would be an easy way to do that.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedure.java:
```diff
@@ -0,0 +1,92 @@
+    private static final String[] PROCEDURE_NAMESPACE = {"db", "fulltext"};
+    private final String identifier;
+    private final FulltextProvider provider;
+    private FulltextProvider.FULLTEXT_INDEX_TYPE type;
```
type can be final

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextFactory.java:
```diff
@@ -0,0 +1,79 @@
+    {
+        this.analyzer = analyzer;
+        this.fileSystem = fileSystem;
+        Factory<IndexWriterConfig> population = () -> IndexWriterConfigs.population( analyzer );
```
still the same as before, why do we use population configuration here?

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextFactory.java:
```diff
@@ -0,0 +1,79 @@
+        this.analyzer = analyzer;
+        this.fileSystem = fileSystem;
+        Factory<IndexWriterConfig> population = () -> IndexWriterConfigs.population( analyzer );
+        partitionFactory = new WritableIndexPartitionFactory( population );
```
usage of `this.fieldName` is inconsistent it should be used everywhere or not used at all

ragadeeshu replied:
Again, used only where needed.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextFactory.java:
```diff
@@ -0,0 +1,79 @@
+    {
+        LuceneIndexStorageBuilder storageBuilder = LuceneIndexStorageBuilder.create();
+        storageBuilder.withFileSystem( fileSystem ).withIndexIdentifier( identifier ).withDirectoryFactory(
+                directoryFactory( false, this.fileSystem ) ).withIndexRootFolder( Paths.get( this.storeDir.getAbsolutePath(), "fulltext" ).toFile() );
```
"fulltext" constant should be extracted

MishaDemianenko replied:
`Paths.get( this.storeDir.getAbsolutePath(), "fulltext" ).toFile()` -> new File(storeDir, "fulltext")

MishaDemianenko replied:
not sure why you need to configure specific directory factory, its already persistent by default

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+    private FulltextProvider( GraphDatabaseService db )
+    {
+        this.db = db;
+        closed = false;
```
non consistent use `this.fieldName`

ragadeeshu replied:
Well, it is used where it is needed, and only where it is needed.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+     * @param db Database used for eventual creation of the provider.
+     * @return
+     */
+    public static synchronized FulltextProvider instance( GraphDatabaseService db )
```
as mentioned above already, it's a lot of additional synch complexity which is still not fully correct, that can be removed

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+                }
+                catch ( IOException e )
+                {
+                    e.printStackTrace();
```
i think we should log those errors instead of printing them

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+
+    Set<WritableFulltext> writableNodeIndices()
+    {
+        return nodeIndices.values().stream().map( WritableFulltext::new ).collect( Collectors.toSet() );
```
since we will do it for every commit, what if we will move forming of those sets to current register thing where we update it with new index as soon as its added and do not touch it anymore

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+
+    String[] getRelationshipProperties()
+    {
+        return relationshipProperties.toArray( new String[0] );
```
same as below

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * Fulltext index type.
+     */
+    public enum FULLTEXT_INDEX_TYPE
```
Please use standard java name conversions for enums

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,166 @@
+    {
+        try
+        {
+            index.maybeRefreshBlocking();
```
It would be really nice if event updater will get some kind of criteria that will determine how often you do this instead. That will provide an ability to test it in an easy way and to make refresh interval configurable

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,166 @@
+        }
+        catch ( IOException e )
+        {
+            e.printStackTrace();
```
thats also should not be here, it should be logged instead

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,166 @@
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( "Unable to update fulltext index", e );
```
not sure exception here will be helpful either since tx will be committed at this point, while you will still get an exception from your transaction which can be super confusing

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,166 @@
+                long entityId = stateEntry.getKey();
+                Map<String,Object> allProperties =
+                        stateEntry.getValue().entrySet().stream().filter( entry -> indexedProperties.contains( entry.getKey() ) ).collect(
+                                Collectors.toMap( entry -> entry.getKey(), entry -> entry.getValue() ) );
```
`entry -> entry.getKey(), entry -> entry.getValue()` can be replaced with method references

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltext.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public int hashCode()
+    {
+        int result = identifier.hashCode();
```
`Objects.hash( identifier, type );` can be used instead of whole method

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltext.java:
```diff
@@ -0,0 +1,133 @@
+
+    private static final String KEY_STATUS = "status";
+    private static final String ONLINE = "online";
+    private static final Map<String,String> ONLINE_COMMIT_USER_DATA = singletonMap( KEY_STATUS, ONLINE );
```
not used

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltext.java:
```diff
@@ -0,0 +1,133 @@
+    private final Analyzer analyzer;
+    private final String identifier;
+    private final FulltextProvider.FULLTEXT_INDEX_TYPE type;
+    private final TaskCoordinator taskCoordinator = new TaskCoordinator( 10, TimeUnit.MILLISECONDS );
```
not used

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltext.java:
```diff
@@ -0,0 +1,133 @@
+    private SimpleFulltextReader createSimpleReader( List<AbstractIndexPartition> partitions ) throws IOException
+    {
+        AbstractIndexPartition singlePartition = getFirstPartition( partitions );
+        return new SimpleFulltextReader( singlePartition.acquireSearcher(), properties.toArray( new String[0] ), analyzer );
```
again another conversion list <-> array

ragadeeshu replied:
This one never goes back to list, though

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/PartitionedFulltextReader.java:
```diff
@@ -0,0 +1,97 @@
+
+    private PrimitiveLongIterator partitionedOperation( Function<ReadOnlyFulltext,PrimitiveLongIterator> readerFunction )
+    {
+        return PrimitiveLongCollections.concat( indexReaders.parallelStream().map( readerFunction::apply ).collect( Collectors.toList() ) );
```
you can use readerFunction directly here, without apply

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/ReadOnlyFulltext.java:
```diff
@@ -0,0 +1,41 @@
+     * @param terms The terms to query for.
+     * @return An iterator over the matching entityIDs, ordered by lucene scoring of the match.
+     */
+    PrimitiveLongIterator query( String... terms );
```
do we really need to distinguish query/fuzzy query on api level for full-text indexes?

ragadeeshu replied:
This is of course subject to change, but for now I want both of them to be readily available. Makes testing easier too :)

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/SimpleFulltextReader.java:
```diff
@@ -0,0 +1,143 @@
+        Query query;
+        try
+        {
+            query = multiFieldQueryParser.parse( String.join( " ", terms ) );
```
can you please explain this fall back mechanics, why do we need it?

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/SimpleFulltextReader.java:
```diff
@@ -0,0 +1,143 @@
+        {
+            DocValuesCollector docValuesCollector = new DocValuesCollector( true );
+            getIndexSearcher().search( query, docValuesCollector );
+            return docValuesCollector.getSortedValuesIterator( NODE_ID_KEY, Sort.RELEVANCE );
```
please be aware that we never used those before, so it can be high chance that those does not work as expected

ragadeeshu replied:
According to the (not all that extensive, but still) tests it seems to work

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/SimpleFulltextReader.java:
```diff
@@ -0,0 +1,143 @@
+    @Override
+    public PrimitiveLongIterator query( String... terms )
+    {
+        MultiFieldQueryParser multiFieldQueryParser = new MultiFieldQueryParser( properties, analyzer );
```
looks like 2 query methods exactly the same, except strings joining with `~` parts so thaose full methods can be combined into one instead, also it's strange that fallback, if that is required does not use any additional foo's for  fuzzy query

MishaDemianenko commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/IndexWriterConfigs.java:
```diff
@@ -108,4 +115,12 @@ public static IndexWriterConfig population()
         return writerConfig;
     }
+
+    public static IndexWriterConfig population( Analyzer analyzer )
```
if we need another population configuration that should delegate most of it's work to default one instead of copy pasting it

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,166 @@
+                Map<String,Object> allProperties = state.get( entityId );
+                if ( allProperties == null || allProperties.isEmpty() )
+                {
+                    index.getIndexWriter().deleteDocuments( LuceneFulltextDocumentStructure.newTermForChangeOrRemove( entityId ) );
```
same point with index writer as above

MishaDemianenko replied:
+1

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedure.java:
```diff
@@ -0,0 +1,92 @@
+{
+    private static final String OUTPUT_NAME = "entityid";
+    private static final String PROCEDURE_NAME = "bloomFulltext";
+    private static final String[] PROCEDURE_NAMESPACE = {"db", "fulltext"};
```
i think Michael/Antons  comment about official API support for procedures is  still valid, would be nice if we use only officially supported/promoted way

MishaDemianenko reviewed (changes requested):
Current state of things is already so much better than before, great step! Still, have some comments here and there

ragadeeshu committed:
Various polish for fulltext addon

ragadeeshu committed:
Add logging for errors when closing fulltext index

ragadeeshu committed:
Code cleanup in fulltext addon

ragadeeshu committed:
Bloom procedures rewritten with procedure API

ragadeeshu committed:
FulltextProvider is no longer singleton

ragadeeshu committed:
Checkstyle fixes in fulltext addon

ragadeeshu committed:
Fulltext addon compilation fixes after rebase

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -0,0 +1,169 @@
+                if ( !allProperties.isEmpty() )
+                {
+                    Document document = LuceneFulltextDocumentStructure.documentRepresentingProperties( entityId, allProperties );
+                    index.getIndexWriter().updateDocument( LuceneFulltextDocumentStructure.newTermForChangeOrRemove( entityId ), document );
```
we can reuse same writer for all of the updates

chrisvest replied:
We also noticed that. I think we'll do it in a different PR because we already have a few changes lined up for this part of the code.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -0,0 +1,177 @@
+        } );
+    }
+
+    synchronized void register( LuceneFulltext fulltextIndex ) throws IOException
```
not sure why this is `synchronized`

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedures.java:
```diff
@@ -0,0 +1,103 @@
+
+    private Stream<Output> queryAsStream( List<String> terms, ReadOnlyFulltext indexReader )
+    {
+        Stream<Output> stream;
```
this can be updated to something like:
```
        PrimitiveLongIterator primitiveLongIterator = indexReader.fuzzyQuery( terms.toArray( new String[0] ) );
        Iterator<Output> iterator = PrimitiveLongCollections.map( Output::new, primitiveLongIterator );
        return StreamSupport.stream( Spliterators.spliteratorUnknownSize( iterator, Spliterator.ORDERED), false );
```

ragadeeshu replied:
Sweet!!

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedures.java:
```diff
@@ -0,0 +1,103 @@
+    @Procedure( name = "db.fulltext.bloomFulltextNodes", mode = READ )
+    public Stream<Output> bloomFulltextNodes( @Name( "terms" ) List<String> terms ) throws Exception
+    {
+        Stream<Output> stream;
```
i would just simply do this (here and method below):
```
        try ( ReadOnlyFulltext indexReader = provider.getReader( BLOOM_NODES, FulltextProvider.FulltextIndexType.NODES ) )
        {
            return queryAsStream( terms, indexReader );
        }
```

MishaDemianenko reviewed (changes requested):
Only a couple of minor comments left! ☕️

ragadeeshu committed:
Small fulltext addon cleanup
